### PR TITLE
Internal Clean Up of Parmest

### PR DESCRIPTION
**Pull request by smartin71**
## Summary/Motivation:

Trying to sort through the issue:
https://github.com/Pyomo/pyomo/issues/3252

## Changes proposed in this PR:

- Removed most of the internal references to "theta" in the parameter estimation package.  
- Referenced pyomo parameter objects directly instead of through names.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/parmest/__init__.py [hunk 1]
```diff
@@ -12,16 +12,16 @@
 from pyomo.common.deprecation import relocated_module_attribute
```
Comment (mrmundt) [inline pyomo/contrib/parmest/__init__.py:12]: It looks like `black` went a little crazy on this file. Can you please revert these changes?

File: pyomo/contrib/parmest/parmest.py [hunk 2]
```diff
@@ -68,10 +68,14 @@
 from pyomo.common.deprecation import deprecated
 from pyomo.common.deprecation import deprecation_warning
 
+from pyomo.core.base.suffix import Suffix
+
+DEPRECATION_VERSION = "6.7.2.dev0"
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:73]: I don't think this is actually used? But if it is being used, it's not the correct value.

File: pyomo/contrib/parmest/parmest.py [hunk 3]
```diff
@@ -68,10 +68,14 @@
...
+
 parmest_available = numpy_available & pandas_available & scipy_available
 
 inverse_reduced_hessian, inverse_reduced_hessian_available = attempt_import(
-    'pyomo.contrib.interior_point.inverse_reduced_hessian'
+    "pyomo.contrib.interior_point.inverse_reduced_hessian"
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:78]: `black` went a little crazy on this file all over the place. Can you please revert those changes? It will make reviewers have to look at a lot more lines that aren't actually any different.

File: pyomo/contrib/parmest/parmest.py [hunk 4]
```diff
@@ -279,20 +299,37 @@ def __init__(
...
+            # (although this is optional for design of experiments)
+            outputs = getattr(model, "experiment_outputs", None)
+            if outputs is None:
+                logger.warning(
+                    "Experiment outputs missing for experiment " + str(i) + "."
+                )
+
+            # check that each model has unknown parameters (not optional)
+            params = getattr(model, "unknown_parameters", None)
+            assert isinstance(params, Suffix)
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:326]: `assert`s like this are problematic because, if the assertion isn't true, the user won't get a helpful message. Can this be changed to something that raises a helpful error message?

File: pyomo/contrib/parmest/parmest.py [hunk 5]
```diff
@@ -687,43 +667,41 @@ def _Q_at_theta(self, thetavals, initialize_parmest_model=False):
...
 
-                for i, theta in enumerate(theta_ref):
-                    # Use parser in ComponentUID to locate the component
-                    var_cuid = ComponentUID(theta)
-                    var_validate = var_cuid.find_component_on(instance)
+                for var_validate in model_param_list:
                     if var_validate is None:
                         logger.warning(
-                            "theta_name %s was not found on the model", (theta)
+                            "param %s was not found on the model", (var_validate.name)
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:686]: May be worth turning this into an `f`-string.
```suggestion
                            f"param {var_validate.name} was not found on the model"
```

File: pyomo/contrib/parmest/parmest.py [hunk 6]
```diff
@@ -844,7 +821,7 @@ def _get_sample_list(self, samplesize, num_samples, replacement=True):
                     if attempts > num_samples:  # arbitrary timeout limit
                         raise RuntimeError(
                             """Internal error: timeout constructing
-                                           a sample, the dim of theta may be too
+                                           a sample, the parameter dimension may be too
                                            close to the samplesize"""
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:825]:
```suggestion
                                           close to the sample size"""
```

File: pyomo/contrib/parmest/parmest.py [hunk 7]
```diff
@@ -1162,72 +1139,63 @@ def objective_at_theta(self, theta_values=None, initialize_parmest_model=False):
...
-                    t.replace("'", "") for t in model_theta_list
-                ], "Theta name {} in 'theta_values' not in 'theta_names' {}".format(
-                    theta_temp, model_theta_list
+            model_names = theta_values.columns
+            # # check if model_names are in model
+            for param in list(model_names):
+                param_temp = param.replace("'", "")  # cleaning quotes from model_names
+                assert param_temp in [
+                    t.replace("'", "") for t in model_param_list
+                ], "Theta name {} in 'theta_values' not in parameter names {}".format(
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1162]: I think this might have been missed because it looks like you were trying to move away from using `theta`?
```suggestion
                ], "Param name {} in 'model_param_list' not in parameter names {}".format(
```

File: pyomo/contrib/parmest/parmest.py [hunk 8]
```diff
@@ -1324,7 +1292,7 @@ def confidence_region_test(
             )
 
         assert isinstance(theta_values, pd.DataFrame)
-        assert distribution in ['Rect', 'MVN', 'KDE']
+        assert distribution in ["Rect", "MVN", "KDE"]
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1295]: There are a few `assert` statements here - see earlier comment about `assert` not always being friendly to the user.

File: pyomo/contrib/parmest/parmest.py [hunk 9]
```diff
@@ -1435,6 +1403,150 @@ def __call__(self, model):
...
+    scenario_name, node_names=None, cb_data=None
+):
+    """
+    This is going to be called by mpi-sppy or the local EF and it will call into
+    the user's model's callback.
+
+    Parameters:
+    -----------
+    scenario_name: `str` Scenario name should end with a number
+    node_names: `None` ( Not used here )
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1416]: Why have it as a parameter if it's not used?

File: pyomo/contrib/parmest/parmest.py [hunk 10]
```diff
@@ -1435,6 +1403,150 @@ def __call__(self, model):
...
+        except TypeError:  # deprecated signature?
+            try:
+                instance = callback(scen_name, node_names)
+            except:
+                print("Failed to create instance using callback; TypeError+")
+                raise
+        except:
+            print("Failed to create instance using callback.")
+            raise
+        """
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1500]: Needed?

File: pyomo/contrib/parmest/parmest.py [hunk 11]
```diff
@@ -1435,6 +1403,150 @@ def __call__(self, model):
...
+
+    if "ThetaVals" in outer_cb_data:
+        thetavals = outer_cb_data["ThetaVals"]
+
+        # dlw august 2018: see mea code for more general theta
+        for name, val in thetavals.items():
+            theta_cuid = ComponentUID(name)
+            theta_object = theta_cuid.find_component_on(instance)
+            if val is not None:
+                # print("Fixing",vstr,"at",str(thetavals[vstr]))
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1541]: Leftover print statement

File: pyomo/contrib/parmest/parmest.py [hunk 12]
```diff
@@ -1435,6 +1403,150 @@ def __call__(self, model):
...
+                theta_object.fix(val)
+            else:
+                # print("Freeing",vstr)
```
Comment (mrmundt) [inline pyomo/contrib/parmest/parmest.py:1544]: Leftover print statement

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 13]
```diff
@@ -22,7 +22,192 @@
...
+        for c in current_keys:
+            if c in comp_map:
+                s[comp_map[c]] = s.pop(c)
+
+        assert len(current_keys) == len(s.keys())
+
+    # print('--- Updated Model ---')
+    # model.pprint()
+    # solver = pyo.SolverFactory('ipopt')
+    # solver.solve(model)
```
Comment (mrmundt) [inline pyomo/contrib/parmest/utils/model_utils.py:204]: Leftover statements?

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 14]
```diff
@@ -22,7 +22,192 @@
...
+
+    return model
+
+
+# deprecated version
+def convert_params_to_vars_deprecated(model, param_names=None, fix_vars=False):
```
Comment (mrmundt) [inline pyomo/contrib/parmest/utils/model_utils.py:210]: Maybe put a deprecation warning over this!

**Review by mrmundt — CHANGES_REQUESTED** [review]

**Comment by blnicho** [comment]
Closing this for now. I'm going to follow-up with Shawn offline.
